In [ ]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report

In [ ]:
spam = pd.read_csv('spam.csv', encoding='latin-1')
spam.columns = ['type', 'text', '', '', '']
spam['type'] = spam['type'].map({'ham': 0, 'spam': 1})
spam.drop(['', '', '', ''], axis=1, inplace=True)
spam.head()

,type,text
0,0,"Go until jurong point, crazy.. Available only ..."
1,0,Ok lar... Joking wif u oni...
2,1,Free entry in 2 a wkly comp to win FA Cup fina...
3,0,U dun say so early hor... U c already then say...
4,0,"Nah I don't think he goes to usf, he lives aro..."


In [ ]:
null_total = spam.isnull().sum()
print(null_total)

type    0
text    0
dtype: int64


In [ ]:
spam.shape

(5572, 2)

In [ ]:
spam.head(100)

,type,text
0,0,"Go until jurong point, crazy.. Available only ..."
1,0,Ok lar... Joking wif u oni...
2,1,Free entry in 2 a wkly comp to win FA Cup fina...
3,0,U dun say so early hor... U c already then say...
4,0,"Nah I don't think he goes to usf, he lives aro..."
...,...,...
95,1,Your free ringtone is waiting to be collected....
96,0,Watching telugu movie..wat abt u?
97,0,i see. When we finish we have loads of loans t...
98,0,Hi. Wk been ok - on hols now! Yes on for a bit...


**TRAIN DATA**

In [ ]:
X = spam['text']
y = spam['type']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=3)

In [ ]:
vectorizer = TfidfVectorizer()
X_train_tfidf = vectorizer.fit_transform(X_train)
X_test_tfidf = vectorizer.transform(X_test)

**LOGISTIC REGRESSION**

In [ ]:
LR = LogisticRegression()
LR.fit(X_train_tfidf, y_train)
y_pred_logistic = LR.predict(X_test_tfidf)

**SVM**

In [ ]:
SVM = SVC()
SVM.fit(X_train_tfidf, y_train)
y_pred_svm = SVM.predict(X_test_tfidf)

**REPORT**

In [ ]:
#Logistic Regression
lr_report = classification_report(y_test, y_pred_logistic)
print(f"Logistic Regression Classification Report: ")
print(lr_report)

accuracy_logistic = accuracy_score(y_test, y_pred_logistic)
print(f"Logistic Regression Accuracy: {accuracy_logistic} \n")

#SVM
svm_report = classification_report(y_test, y_pred_svm)
print(f"SVM Classification Report: ")
print(svm_report)

accuracy_svm = accuracy_score(y_test, y_pred_svm)
print(f"SVM Accuracy: {accuracy_svm}")

Logistic Regression Classification Report: 
              precision    recall  f1-score   support

           0       0.97      1.00      0.99       960
           1       1.00      0.83      0.90       155

    accuracy                           0.98      1115
   macro avg       0.99      0.91      0.95      1115
weighted avg       0.98      0.98      0.97      1115

Logistic Regression Accuracy: 0.9757847533632287 

SVM Classification Report: 
              precision    recall  f1-score   support

           0       0.98      1.00      0.99       960
           1       1.00      0.90      0.95       155

    accuracy                           0.99      1115
   macro avg       0.99      0.95      0.97      1115
weighted avg       0.99      0.99      0.99      1115

SVM Accuracy: 0.9865470852017937


**Difference of Performance**

*Overall Accuracy*
*   SVM slightly outperforms Logistic Regression (98.65% vs 97.57%)

*Spam Detection*
*   SVM has better recall (0.90 vs 0.83) → which means it catches more spam
*   SVM has better F1-score for spam (0.95 vs 0.90) → which means SVM has a better balance between: Precision (correctly identifying actual spam) and
Recall (catching as many spam emails as possible)



